### PR TITLE
Grouped GEMM skip useless computation for unaligned Ms

### PR DESCRIPTION
**Pull request by yukuai26**
Skip useless computation on M and reconstruct test_m_grouped_gemm_contiguous and test_m_grouped_gemm_masked'.

In test_m_grouped_gemm_contiguous, there is a speedup of 0-15 TFLOPS (Due to the larger expected_m_per_group, the speedup effect is not very significant).

In test_m_grouped_gemm_masked, there is a speedup of 10-20 TFLOPS.

For GEMM, incorporating this optimization may affect the efficiency of certain shapes. I tested with m = 4160 and observed some speedup, but when m = 4096, there is a 10-20 TFLOPS reduction in efficiency.

This scenario can be specifically optimized by adding a template parameter to check whether m is a multiple of block_m. However, since GEMM is not the current optimization focus, I have not implemented this enhancement yet.

### CODE REVIEW

**Comment by CalebDu** [comment]
@LyricZhao I try to use m_indices=-1 to skip padding M block computation in https://github.com/vllm-project/vllm/pull/17934#issuecomment-2980367911. In noEP inference, vllm will pre-allocate bulk of memory(m-dim larger than actual n_token*topk beacause of padding  block to 128 for each expert) for permuted hidden in  group gemm. But I encounter program hang in vllm moe test for deepseekv3. And I try to build simulated test in deep gemm test as following code.
``` python
def construct_contiguous_grouped(num_groups: int, expected_m_per_group: int, k: int, n: int) -> \
        Tuple[int, Tuple[torch.Tensor, torch.Tensor], Tuple[torch.Tensor, torch.Tensor], torch.Tensor, torch.Tensor, torch.Tensor]:
    alignment = get_m_alignment_for_contiguous_layout()
    group_ms = [int(expected_m_per_group * random.uniform(0.7, 1.3)) for _ in range(num_groups)]
    actual_m = sum([ceil_div(x, alignment) * alignment for x in group_ms])
    # padding some dummy data
    padding_m = actual_m + 1024*10 
    x = torch.randn((padding_m, k), device='cuda', dtype=torch.bfloat16)
    y = torch.randn((num_groups, n, k), device='cuda', dtype=torch.bfloat16)    
    m_indices = torch.empty(padding_m, device='cuda', dtype=torch.int32).fill_(-1)
```
I run contiguous group gemm test with padding dummy data and the test program also hang without termination. I notice that producer warp still load dummy data from global memory to shared memory with triggering mbarrier `full.arrive_expect_tx()/ empty.wait()`and consumer warp group sets `kNumInnerStages`=0 with `SkipComputation` without triggering mbarrier `full.wait()/ empty.arrive()`. I am not sure whether this implementation will trigger potiential mbarrier dead lock.

**Comment by BBuf** [comment]
> @LyricZhao I try to use m_indices=-1 to skip padding M block computation in [vllm-project/vllm#17934 (comment)](https://github.com/vllm-project/vllm/pull/17934#issuecomment-2980367911). In noEP inference, vllm will pre-allocate bulk of memory(m-dim larger than actual n_token*topk beacause of padding block to 128 for each expert) for permuted hidden in group gemm. But I encounter program hang in vllm moe test for deepseekv3. And I try to build simulated test in deep gemm test as following code.
> 
> ```python
> def construct_contiguous_grouped(num_groups: int, expected_m_per_group: int, k: int, n: int) -> \
>         Tuple[int, Tuple[torch.Tensor, torch.Tensor], Tuple[torch.Tensor, torch.Tensor], torch.Tensor, torch.Tensor, torch.Tensor]:
>     alignment = get_m_alignment_for_contiguous_layout()
>     group_ms = [int(expected_m_per_group * random.uniform(0.7, 1.3)) for _ in range(num_groups)]
>     actual_m = sum([ceil_div(x, alignment) * alignment for x in group_ms])
>     # padding some dummy data
>     padding_m = actual_m + 1024*10 
>     x = torch.randn((padding_m, k), device='cuda', dtype=torch.bfloat16)
>     y = torch.randn((num_groups, n, k), device='cuda', dtype=torch.bfloat16)    
>     m_indices = torch.empty(padding_m, device='cuda', dtype=torch.int32).fill_(-1)
> ```
> 
> I run contiguous group gemm test with padding dummy data and the test program also hang without termination. I notice that producer warp still load dummy data from global memory to shared memory with triggering mbarrier `full.arrive_expect_tx()/ empty.wait()`and consumer warp group sets `kNumInnerStages`=0 with `SkipComputation` without triggering mbarrier `full.wait()/ empty.arrive()`. I am not sure whether this implementation will trigger potiential mbarrier dead lock.

I also reproduced this hanging issue on h200. May I ask when this can be fixed? Thank you. @LyricZhao

**Comment by LyricZhao** [comment]
Thanks for feedback! I will fix them in today.

**Comment by LyricZhao** [comment]
Sorry, I didn't reproduce the bug using the latest commit and the test below:

```python
def construct_contiguous_grouped(num_groups: int, expected_m_per_group: int, k: int, n: int) -> \
        Tuple[int, Tuple[torch.Tensor, torch.Tensor], Tuple[torch.Tensor, torch.Tensor], torch.Tensor, torch.Tensor, torch.Tensor]:
    alignment = get_m_alignment_for_contiguous_layout()
    group_ms = [int(expected_m_per_group * random.uniform(0.7, 1.3)) for _ in range(num_groups)]
    actual_m = sum([ceil_div(x, alignment) * alignment for x in group_ms])
    # padding some dummy data
    padding_m = actual_m + 1024*10
    x = torch.randn((padding_m, k), device='cuda', dtype=torch.bfloat16)
    y = torch.randn((num_groups, n, k), device='cuda', dtype=torch.bfloat16)
    m_indices = torch.empty(padding_m, device='cuda', dtype=torch.int32).fill_(-1)
    out = torch.empty((padding_m, n), device='cuda', dtype=torch.bfloat16)
    ref_out = torch.randn((padding_m, n), device='cuda', dtype=torch.bfloat16)

    start = 0
    for i, group_m in enumerate(group_ms):
        actual_end = start + group_m
        aligned_end = start + ceil_div(group_m, alignment) * alignment
        m_indices[start:actual_end] = i
        m_indices[actual_end:aligned_end] = -1
        ref_out[start:aligned_end] = x[start:aligned_end] @ y[i].t()
        start = aligned_end
    ref_out = torch.where((m_indices == -1).unsqueeze(1), torch.zeros_like(ref_out), ref_out)

    assert padding_m % 4 == 0, f'TMA alignment error: {padding_m}'
    x_fp8 = per_token_cast_to_fp8(x)
    y_fp8 = (torch.empty_like(y, dtype=torch.float8_e4m3fn), torch.empty((num_groups, ceil_div(n, 128), k // 128), device='cuda', dtype=torch.float))
    for i in range(num_groups):
        y_fp8[0][i], y_fp8[1][i] = per_block_cast_to_fp8(y[i])

    return padding_m, x_fp8, y_fp8, m_indices, out, ref_out
```

Can you please give me the full test script or dump the tensor?


> I run contiguous group gemm test with padding dummy data and the test program also hang without termination. I notice that producer warp still load dummy data from global memory to shared memory with triggering mbarrier full.arrive_expect_tx()/ empty.wait()and consumer warp group sets kNumInnerStages=0 with SkipComputation without triggering mbarrier full.wait()/ empty.arrive(). I am not sure whether this implementation will trigger potiential mbarrier dead lock.

The unaligned cases will trigger full wait and empty arrive:

```cuda
// Wait unaligned cases
  #pragma unroll
  for (uint32_t s = kNumInnerStages; s < kNumStages; ++ s) {
      full_barriers[s]->wait((scheduler.current_iter * kNumIterations + k_iter) & 1);
      empty_barrier_arrive(s);
  }
```

**Comment by CalebDu** [comment]
> Sorry, I didn't reproduce the bug using the latest commit and the test below:
> 
> ```python
> def construct_contiguous_grouped(num_groups: int, expected_m_per_group: int, k: int, n: int) -> \
>         Tuple[int, Tuple[torch.Tensor, torch.Tensor], Tuple[torch.Tensor, torch.Tensor], torch.Tensor, torch.Tensor, torch.Tensor]:
>     alignment = get_m_alignment_for_contiguous_layout()
>     group_ms = [int(expected_m_per_group * random.uniform(0.7, 1.3)) for _ in range(num_groups)]
>     actual_m = sum([ceil_div(x, alignment) * alignment for x in group_ms])
>     # padding some dummy data
>     padding_m = actual_m + 1024*10
>     x = torch.randn((padding_m, k), device='cuda', dtype=torch.bfloat16)
>     y = torch.randn((num_groups, n, k), device='cuda', dtype=torch.bfloat16)
>     m_indices = torch.empty(padding_m, device='cuda', dtype=torch.int32).fill_(-1)
>     out = torch.empty((padding_m, n), device='cuda', dtype=torch.bfloat16)
>     ref_out = torch.randn((padding_m, n), device='cuda', dtype=torch.bfloat16)
> 
>     start = 0
>     for i, group_m in enumerate(group_ms):
>         actual_end = start + group_m
>         aligned_end = start + ceil_div(group_m, alignment) * alignment
>         m_indices[start:actual_end] = i
>         m_indices[actual_end:aligned_end] = -1
>         ref_out[start:aligned_end] = x[start:aligned_end] @ y[i].t()
>         start = aligned_end
>     ref_out = torch.where((m_indices == -1).unsqueeze(1), torch.zeros_like(ref_out), ref_out)
> 
>     assert padding_m % 4 == 0, f'TMA alignment error: {padding_m}'
>     x_fp8 = per_token_cast_to_fp8(x)
>     y_fp8 = (torch.empty_like(y, dtype=torch.float8_e4m3fn), torch.empty((num_groups, ceil_div(n, 128), k // 128), device='cuda', dtype=torch.float))
>     for i in range(num_groups):
>         y_fp8[0][i], y_fp8[1][i] = per_block_cast_to_fp8(y[i])
> 
>     return padding_m, x_fp8, y_fp8, m_indices, out, ref_out
> ```
> 
> Can you please give me the full test script or dump the tensor?
> 
> > I run contiguous group gemm test with padding dummy data and the test program also hang without termination. I notice that producer warp still load dummy data from global memory to shared memory with triggering mbarrier full.arrive_expect_tx()/ empty.wait()and consumer warp group sets kNumInnerStages=0 with SkipComputation without triggering mbarrier full.wait()/ empty.arrive(). I am not sure whether this implementation will trigger potiential mbarrier dead lock.
> 
> The unaligned cases will trigger full wait and empty arrive:
> 
> ```cuda-c++
> // Wait unaligned cases
>   #pragma unroll
>   for (uint32_t s = kNumInnerStages; s < kNumStages; ++ s) {
>       full_barriers[s]->wait((scheduler.current_iter * kNumIterations + k_iter) & 1);
>       empty_barrier_arrive(s);
>   }
> ```

I lost a detail. I can reproduce the bug with setting group=256 in test for no EP inference.
https://github.com/deepseek-ai/DeepGEMM/blob/0c88cd01392c1073c7049a97d6328c7bba9b3947/tests/test_core.py#L193C1-L195C98

**Comment by LyricZhao** [comment]
@yukuai26 will follow this issue up.

**Comment by yukuai26** [comment]
@CalebDu
Sorry, I didn't reproduce the bug using the latest commit and the test below:
```python
def construct_contiguous_grouped(num_groups: int, expected_m_per_group: int, k: int, n: int) -> \
        Tuple[int, Tuple[torch.Tensor, torch.Tensor], Tuple[torch.Tensor, torch.Tensor], torch.Tensor, torch.Tensor, torch.Tensor]:
    alignment = get_m_alignment_for_contiguous_layout()
    group_ms = [int(expected_m_per_group * random.uniform(0.7, 1.3)) for _ in range(num_groups)]
    actual_m = sum([ceil_div(x, alignment) * alignment for x in group_ms])
    # padding some dummy data
    padding_m = actual_m + 1024*10 
    x = torch.randn((padding_m, k), device='cuda', dtype=torch.bfloat16)
    y = torch.randn((num_groups, n, k), device='cuda', dtype=torch.bfloat16)    
    m_indices = torch.empty(padding_m, device='cuda', dtype=torch.int32).fill_(-1)
    out = torch.empty((padding_m, n), device='cuda', dtype=torch.bfloat16)
    ref_out = torch.randn((padding_m, n), device='cuda', dtype=torch.bfloat16)

    start = 0
    for i, group_m in enumerate(group_ms):
        actual_end = start + group_m
        aligned_end = start + ceil_div(group_m, alignment) * alignment
        m_indices[start:actual_end] = i
        m_indices[actual_end:aligned_end] = -1
        ref_out[start:aligned_end] = x[start:aligned_end] @ y[i].t()
        start = aligned_end
    ref_out = torch.where((m_indices == -1).unsqueeze(1), torch.zeros_like(ref_out), ref_out)

    assert padding_m % 4 == 0, f'TMA alignment error: {padding_m}'
    x_fp8 = per_token_cast_to_fp8(x)
    y_fp8 = (torch.empty_like(y, dtype=torch.float8_e4m3fn), torch.empty((num_groups, ceil_div(n, 128), k // 128), device='cuda', dtype=torch.float))
    for i in range(num_groups):
        y_fp8[0][i], y_fp8[1][i] = per_block_cast_to_fp8(y[i])

    return padding_m, x_fp8, y_fp8, m_indices, out, ref_out
```
```python
    for num_groups, expected_m_per_group, k, n in ((256, 16, 7168, 4096), (256, 128, 2048, 7168),
                                                   (256, 64, 7168, 4096), (256, 128, 2048, 7168),
                                                   (256, 32, 7168, 4096), (256, 128, 2048, 7168)):
```

Can you please give me the full test script or dump the tensor?

If the build process is complex, please feel free to add me on WeChat (yukuai26_) and let me know how to reproduce the error. Thank you!

**Comment by yizhang2077** [comment]
hi @yukuai26  I did unit test with mask group gemm (using qwen3-235b-fp8 shape), while I also reproduced stable hang case in my h100/h200 env (before this commit hang case will not happen). This is the code for reproduction:
``` python
def test_m_grouped_gemm_masked() -> None:                                                                               
    print('Testing grouped masked GEMM:')                                                                                                                                                                                                        
    for num_groups, expected_m_per_group,k,n in ((16, 114, 4096,3072), (16, 114, 1536,4096)):                           
        x_fp8, y_fp8, masked_m, out, ref_out = construct_masked_grouped(num_groups, 4096, expected_m_per_group, k, n)                                       
        def test_func():                                                                                            
            deep_gemm.m_grouped_gemm_fp8_fp8_bf16_nt_masked(x_fp8, y_fp8, out, masked_m, expected_m_per_group)                                                                                                             
            # Test performance with fixed shapes                                                                        
            # noinspection PyUnboundLocalVariable                                                                       
        valid_m = masked_m.sum().item()                                                                             
        t = bench_kineto(test_func, 'fp8_gemm', suppress_kineto_output=True)                                        
        print(f' > Perf ({num_groups=}, expected_m_per_group={expected_m_per_group:4}, n={n:4}, k={k:4}): {t * 1e6:4.0f} us | '
                  f'throughput: {2 * valid_m * n * k / t / 1e12:4.0f} TFLOPS, '                                         
                  f'{(valid_m * k + num_groups * k * n + valid_m * n * 2) / 1e9 / t:4.0f} GB/s')                        

test_m_grouped_gemm_masked()
```

**Comment by MARD1NO** [comment]
Hi, I also encounter deepgemm hang when using `m_grouped_gemm_fp8_fp8_bf16_nt_masked`, and this commit `391755ada0ffefa9a6a52b6f14dcaf22d1a463e0` will not happen hang

**Comment by LyricZhao** [comment]
@yukuai26 please follow this up.

**Comment by yukuai26** [comment]
I'm sorry, but I tried running the program you provided on H800 for two days, and I still cannot reproduce your hang case. Below is my test program. Could you please dump a tensor file that can cause the hang, or provide another way to reproduce the issue?
``` python
def test_m_grouped_gemm_masked() -> None:                                                                               
    print('Testing grouped masked GEMM:')
    dump_dir = "tensor_dumps"
    import os
    os.makedirs(dump_dir, exist_ok=True)
    for i in range(1000000000):                                                                                                                                                                                                        
        for num_groups, expected_m_per_group,k,n in ((16, 114, 4096,3072), (16, 114, 1536,4096)):                           
            x_fp8, y_fp8, masked_m, out, ref_out = construct_masked_grouped(num_groups, 4096, expected_m_per_group, k, n)                                       
            def test_func():                                                                                            
                deep_gemm.m_grouped_gemm_fp8_fp8_bf16_nt_masked(x_fp8, y_fp8, out, masked_m, expected_m_per_group)                                                                                                             
                # Test performance with fixed shapes                                                                        
                # noinspection PyUnboundLocalVariable

            filename = f"tensors_iter{i % 2}_groups{num_groups}_m{expected_m_per_group}_k{k}_n{n}.pt"
            filepath = os.path.join(dump_dir, filename)
            
            tensor_dict = {
                'x_fp8': x_fp8,
                'y_fp8': y_fp8,
                'out': out,
                'masked_m': masked_m,
                'expected_m_per_group': expected_m_per_group,
                'metadata': {
                    'num_groups': num_groups,
                    'k': k,
                    'n': n,
                    'iteration': i
                }
            }
            
            torch.save(tensor_dict, filepath)
            print(f" > Dumped tensors to: {filepath}")                                                                       
            valid_m = masked_m.sum().item()                                                                             
            t = bench_kineto(test_func, 'fp8_gemm', suppress_kineto_output=True)                           
            print(f' > Perf ({num_groups=}, expected_m_per_group={expected_m_per_group:4}, n={n:4}, k={k:4}): {t * 1e6:4.0f} us | '
                    f'throughput: {2 * valid_m * n * k / t / 1e12:4.0f} TFLOPS, '                                         
                    f'{(valid_m * k + num_groups * k * n + valid_m * n * 2) / 1e9 / t:4.0f} GB/s')  
···

**Comment by Bruce-Lee-LY** [comment]
@yukuai26 @CalebDu @BBuf @yizhang2077 same hang in grouped gemm contiguous, may be related to the environment? As long as the value of the skipped part of m_indices is set to -1, it can be reproduced. Otherwise, the result will be normal.
Commit: 0c88cd0
GPU: H20
CUDA: 12.6
CUDA Driver: 550.144.04
